zkevm-circuits/src/evm_circuit.rs

```diff
@@ -1 +1,916 @@
+                    cell.assign(region, offset, Some(F::from_u64(*byte as u64)))
+                })
+                .collect()
+        })
+    }
+}
+
+impl<F: FieldExt> Expr<F> for Word<F> {
+    fn expr(&self) -> Expression<F> {
+        self.expression.clone()
+    }
+}
+
+// TODO: Should we maintain these state in circuit or we prepare all of them
+// when parsing debug trace?
+#[derive(Debug)]
+pub(crate) struct CoreStateInstance {
+    is_executing: bool,
+    global_counter: usize,
+    call_id: usize,
+    program_counter: usize,
+    stack_pointer: usize,
+    gas_counter: usize,
+}
+
+impl CoreStateInstance {
+    fn new(call_id: usize) -> Self {
+        Self {
+            is_executing: false,
+            global_counter: 1,
+            call_id,
+            program_counter: 0,
+            stack_pointer: 1024,
+            gas_counter: 0,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+enum Case {
+    Success,
+    // out of gas
+    OutOfGas,
+    // call depth exceeded 1024
+    DepthOverflow,
+    // insufficient balance for transfer
+    InsufficientBalance,
+    // contract address collision
+    ContractAddressCollision,
+    // execution reverted
+    Reverted,
+    // max code size exceeded
+    MaxCodeSizeExceeded,
+    // invalid jump destination
+    InvalidJump,
+    // static call protection
+    WriteProtection,
+    // return data out of bound
+    ReturnDataOutOfBounds,
+    // contract must not begin with 0xef due to EIP #3541 EVM Object Format
+    // (EOF)
+    InvalidBeginningCode,
+    // stack underflow
+    StackUnderflow,
+    // stack overflow
+    StackOverflow,
+    // invalid opcode
+    InvalidCode,
+}
+
+// TODO: use ExecutionStep from bus_mapping
+pub(crate) struct ExecutionStep {
+    opcode: OpcodeId,
+    case: Case,
+    values: Vec<[u8; 32]>,
+}
+
+// TODO: use Operation from bus_mapping
+pub(crate) struct Operation<F> {
+    gc: usize,
+    target: Target,
+    is_write: bool,
+    values: [F; 4],
+}
+
+#[derive(Clone)]
+struct EvmCircuit<F> {
+    q_step: Selector,
+    qs_byte_lookup: Column<Advice>,
+    fixed_table: [Column<Fixed>; 4],
+    rw_table: [Column<Advice>; 7],
+    op_execution_gadget: OpExecutionGadget<F>,
+}
+
+impl<F: FieldExt> EvmCircuit<F> {
+    fn configure(meta: &mut ConstraintSystem<F>, r: F) -> Self {
+        let q_step = meta.selector();
+        let qs_byte_lookup = meta.advice_column();
+        let advices = (0..CIRCUIT_WIDTH)
+            .map(|_| meta.advice_column())
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap();
+
+        // TODO: consider encode first 3 information into one field element
+        // rw_table contains read-write data in bus mapping for lookup
+        let rw_table = [
+            meta.advice_column(), // global_counter
+            meta.advice_column(), // target
+            meta.advice_column(), // is_write
+            meta.advice_column(), // val1
```

Comment:
Add comment for actual meaning.

```suggestion
            meta.advice_column(), // val1, call_id
```

Reply:
These `val*` mean different things depends on the target, so I left it arbitrary instead of concrete meaning. For stack or memroy it would be call_id, but for storage it would be account address. Will try to document it in a better way to clarify.